Add rate limiter for peer messages

Gmin2 opened the pull request:
Adds rate limiting for incoming Workbase, UserWorkbase, MiningShare, Inventory and transactions messages to prevent DoS attacks and added a test

![Screenshot from 2025-03-19 02-03-08](https://github.com/user-attachments/assets/9de8f779-809c-460a-981c-eb86f0222038)

Fixes #30

pool2win commented on src/node/mod.rs:
```diff
@@ -219,6 +229,23 @@ impl Node {
                     Ok(())
                 }
                 P2PoolBehaviourEvent::Gossipsub(gossip_event) => {
+                    if let gossipsub::Event::Message {
```
We seem to be repeating this code structure to check for rate limit and then disconnect. Is it possible to abstract it out it in a function, so we don't repeat ourselves?

pool2win commented on src/node/rate_limiter.rs:
```diff
@@ -0,0 +1,121 @@
+use std::time::{Duration, Instant};
+use tracing::warn;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
```
Was it not possible to directly using the messages::Message type here? By introducing another type do we make it harder to keep things consistent over the long term?

Gmin2 replied:
i think it will be better to use the new type to avoid accidently applying a rate limit to a new message type added in the future

pool2win commented on src/node/rate_limiter.rs:
```diff
@@ -0,0 +1,121 @@
+use tracing::warn;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum MessageType {
```
If we do keep the type, lets add a doc string explaining the need for this and how the type will be used.

pool2win commented on src/node/rate_limiter.rs:
```diff
@@ -0,0 +1,121 @@
+    }
+}
+
+#[derive(Debug, Clone)]
```
Add docstring.

pool2win commented on src/node/rate_limiter.rs:
```diff
@@ -0,0 +1,121 @@
+    }
+}
+
+#[derive(Debug, Clone, Default)]
```
Add docstring.

pool2win commented on src/node/rate_limiter.rs:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    pub fn check_rate_limit(
```
Add docstring.

pool2win commented on src/node/rate_limiter.rs:
```diff
@@ -0,0 +1,121 @@
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct RateLimiter {
```
The rate limiter will end up being called from multiple async tasks as we introduce parallel processing of events. We should consider how to make this implemention of RateLimiter thread safe.

pool2win reviewed (changes requested):
Nice one! Thanks for taking the time to work on this.

There's a few nits about adding docstrings. It helps others coming to the code base and to our future selves too if there some documenation explaining the need for the structs/functions and how they should be used.

Also, we should put in a little more work to make the RateLimiter threadsafe. Maybe we can get away with using a mutex for now. Let me know what you think.

pool2win commented:
We also need to provide tests for the new module and it's use, so we know it works as expected. :)

Gmin2 commented:
push the new change by implementing a new rate limiter using `sliding window` technique and make it thread safe by using `tokio::Mutex`, also added the unit test for it  and added the docstring wherever yu have requested

pool2win commented:
Can you please rebase onto main. The codecov result uploads consistently fail and I realised upgrading to codecov action v4 resolves the problem. The main branch has the fix now, so if you rebase, it will make it easier to get that codecov action to finish. I have already tried running it a few time. Hope it is not too much of a bother. Thanks.

Gmin2 commented:
Done , but the test coverage is below the expected target , so should i add test covering it ?

cc @pool2win

pool2win commented:
> Done , but the test coverage is below the expected target , so should i add test covering it ?
> 
> cc @pool2win

It is OK that the node module coverage fell a bit, I can take care of that. We should definitely try and increase the coverage for the rate limiter module.

I was having another think about the MessageType type. It has a few code smells:

1. Its name ends with "Type", which is kind of unnecessary. Then if you drop Type, you end up with a type called Message, which we already have available from an imported module.
2. The From trait is simply converting the type.

I do think we should drop MessageType and directly use `crate::node::messages::Message`. It will be nice to merge this PR, so let's make it happen!

Gmin2 commented:
> I do think we should drop MessageType and directly use `crate::node::messages::Message`. It will be nice to merge this PR, so let's make it happen!

Done
